wrappers/wrapper.rb

```diff
@@ -739,87 +767,108 @@ def detect_unfeasible_services(vrp)
-                earliest_arrival += matrix[:time][vehicle.start_point.matrix_index][index] if vehicle.start_point_id
-                timely_arrival_not_possible = activity.timewindows.all?{ |tw|
-                  tw.end && earliest_arrival && earliest_arrival > tw.end
-                }
-              end
+      if !activity.late_multiplier&.positive? # if service tw_violation is not allowed
+        earliest_arrival = 0
+        earliest_arrival += vehicle.timewindow.start                                       if vehicle.timewindow&.start
+        earliest_arrival += matrix[:time][vehicle.start_point.matrix_index][service_index] if vehicle.start_point_id
+        timely_arrival_not_possible = activity.timewindows.all?{ |tw|
+          tw.end && earliest_arrival && earliest_arrival > tw.end
+        }
+      end
 
-              if !vehicle.cost_late_multiplier&.positive? && (vehicle.timewindow&.end || vehicle.sequence_timewindows&.size&.positive?) # if vehicle tw_violation is not allowed
-                vehicle_end = vehicle.timewindow&.end || vehicle.sequence_timewindows.collect(&:end).max
-                latest_arrival = vehicle_end - activity.duration
-                latest_arrival -= matrix[:time][index][vehicle.end_point.matrix_index] if vehicle.end_point_id
-                timely_return_not_possible = activity.timewindows.all?{ |tw|
-                  latest_arrival && latest_arrival < tw.start
-                }
-              end
-              timely_arrival_not_possible || timely_return_not_possible
-            }
-          }
+      if !vehicle.cost_late_multiplier&.positive? && (vehicle.timewindow&.end || vehicle.sequence_timewindows&.size&.positive?) # if vehicle tw_violation is not allowed
+        vehicle_end = vehicle.timewindow&.end || vehicle.sequence_timewindows.collect(&:end).max
+        latest_arrival = vehicle_end - activity.duration
+        latest_arrival -= matrix[:time][service_index][vehicle.end_point.matrix_index] if vehicle.end_point_id
+        timely_return_not_possible = activity.timewindows.all?{ |tw|
+          latest_arrival && latest_arrival < tw.start
+        }
+      end
+      timely_arrival_not_possible || timely_return_not_possible
+    end
+
+    def check_distances(vrp, unfeasible)
+      unfeasible = check(vrp, :time, unfeasible)
+      unfeasible = check(vrp, :distance, unfeasible)
+      unfeasible = check(vrp, :value, unfeasible)
+
+      vrp.services.each{ |service|
+        if !service_reachable_by_any_vehicle(vrp, service)
+          add_unassigned(
+            unfeasible, vrp, service, 'Service cannot be served due to vehicle parameters -- e.g., timewindow, distance limit, etc.')
+          next
         end
 
-        add_unassigned(unfeasible, vrp, service, 'Service cannot be reached within its timewindows') if service_unreachable_within_its_tw
+        next unless service_unreachable_within_its_tw(vrp, service)
```

Comment:
Why don't we reorder this logic like you did in `reject_services_from_inconsistent_relations` code so that timewindow logic is called only for "compatible" vehicles? (This is less ideal than the [:compatible_vehicles] idea that we first tried but it doesn't need more dev, even if deletes extra code.)

``` ruby
              (service.activity ? [service.activity] : service.activites).any?{ |activity|
                service_reachable_by_vehicle(vrp, activity, vehicle) &&
                  service_uncompatible_with_vehicle_timewindows(vrp, activity, vehicle) 
```

Then we can use a more encompassing reason like `There is no vehicle (with compatible skills, timewindows, duration, depot distance, etc.) that can serve this service`.